File: src_py/utils/transport.py
from typing import List, Dict, Tuple
import igraph as ig
import numpy as np


class Input:
    def __init__(
            self,
            graph: ig.Graph,
            weights: np.ndarray,
            ndds: List[int],
            cycle_length: int,
            chain_length: int,
            forbidden_nodes: List[int] = None
            ):
        self.graph = graph
        self.weights = weights
        self.ndds = ndds
        self.cycle_length = cycle_length
        self.chain_length = chain_length
        self.solver_instance = None
        self.start_time = None

        self.forbidden_nodes = forbidden_nodes


class Output:
    def __init__(
            self,
            match_edges: Dict[Tuple[int], float],
            match_cycles: {Dict[int, float], None},
            graph_cycles: {List[Tuple[int]], None},
            value: float,
            optimal: bool,
            ):
        self.match_edges = match_edges
        self.match_cycles = match_cycles
        self.graph_cycles = graph_cycles
        self.value = value
        self.optimal = optimal
        self.time = -1

File: src_jl/transport.jl
mutable struct Input
    graph::DiGraph
    weights::Array{Float64,2}
    ndds::Array{Int64,1}
    max_cycle_length::Int64
    max_chain_length::Int64
    solver_instance::Any
    time_param::String
    time_factor::Float64
    start_time::Float64

    function Input(
        graph::DiGraph,
        weights::Array{Float64,2},
        ndds::Array{Int64,1},
        max_cycle_length::Int64,
        max_chain_length::Int64,
        solver_instance::Any,
        time_param::String,
        time_factor::Float64,
        start_time::Float64
    )
        new(graph, weights, ndds, max_cycle_length, max_chain_length, solver_instance, time_param, time_factor, start_time)
    end
end

mutable struct Output
    match_vertices::Array{Float64,1}
    match_edges::Any
    match_cycles::Union{Array{Float64,1}, Nothing}
    graph_cycles::Union{Array{Array{Int64,1},1}, Nothing}
    value::Float64
    optimal::Bool

    function Output()
        match_vertices = Float64[]
        match_edges = Float64[]
        match_cycles = Array{Float64}[]
        graph_cycles = Array{Float64}[]
        value = typemin(Float64)
        optimal = false
        new(match_vertices, match_edges, match_cycles, graph_cycles, value, optimal)
    end

    function Output(
        match_vertices::Array{Float64,1},
        match_edges::Any,
        match_cycles::Union{Array{Float64,1}, Nothing},
        graph_cycles::Union{Array{Array{Int64,1},1}, Nothing},
        value::Float64,
        optimal::Bool,
    )
        new(match_vertices, match_edges, match_cycles, graph_cycles, value, optimal)
    end
end
